replica/table.cc

```diff
@@ -550,9 +550,27 @@ void table::enable_off_strategy_trigger() {
 
 storage_group_manager::~storage_group_manager() = default;
 
+future<> storage_group_manager::parallel_foreach_storage_group(std::function<future<>(storage_group&)> f) {
+    auto storage_groups = _storage_groups;
```

Comment:
This isn't required since coroutine::parallel_for_each does not preempt when invoking the function on each element (same for legacy parallel_for_each).
It may yield only when co_awaiting the resulting futures.

That said, it's not wrong either.

Reply:
you're right, but it's better to be explicit here right?

Reply:
It's a short list and just a copy of lw_shared_ptr so it shouldn't be terrible.
Does it make sense to use boost::copy_ramge to copy the shared group ptrs into a linear vector rather than copying the flat hash map?

Reply:
Doesn't the caller care that its callback was executed for a storage group that was removed?

Reply:
> Doesn't the caller care that its callback was executed for a storage group that was removed?

I don't think so. such procedures are used for table wide ops (e.g. nodetool flush), and if we have a tablet being migrated out, it's ok that this table wide op skips a storage group that was removed.

e.g.

1) tablet is being migrated out
2) runs a major compact / flush
3) loop takes a copy of storage map
4) cleanup removes storage for the tablet
5) loop reaches the storage group for that tablet and skips it because it no longer belongs to this node

do you think we need to enforce it somehow other than documentation?

Reply:
it's a trade off I think we're willing to make, to not block migrations / topology changes by taking a lock in cleanup, or update_effective_replication_map() (which keeps old erm alive till completion).

Reply:
I think it's okay.
 - nodetools-style operations are inherently racy wrt migration
 - more global operations should use cluster-wide methods to prevent the race

Reply:
I add doc to compaction_group.hh explaining the approach and scenarios to be handled.

Reply:
added*